examples/tests/invalid/json.w

```diff
@@ -18,23 +17,81 @@ let a: Array<str> = j;
 // Immutable Json
 let foreverJson = Json {a: "hello"};
 foreverJson.set("a", "world!");
-//           ^^^ Unknown symbol "set" (TODO: better error message https://github.com/winglang/wing/issues/1660) 
+//          ^^^ Unknown symbol "set" (TODO: better error message https://github.com/winglang/wing/issues/1660) 
 
 let bkt = new cloud.Bucket();
 let jArr = Json [bkt];
-//                ^^^ Expected "Json" elements to be Json Value (https://www.json.org/json-en.html), but got "Bucket" which is not Json Value
+//               ^^^ Expected "Json" elements to be Json Value (https://www.json.org/json-en.html), but got "Bucket" which is not Json Value
 
 let jsonObj = Json { boom: bkt };
-//                          ^^^ Expected "Json" elements to be Json Value (https://www.json.org/json-en.html), but got "Bucket" which is not Json Value
+//                         ^^^ Expected "Json" elements to be Json Value (https://www.json.org/json-en.html), but got "Bucket" which is not Json Value
 
 let jsonIncomplete = Json;
-//                    ^^^^ Expected a "Json" element
+//                   ^^^^ Expected a "Json" element
 
 let tryNum: num? = j.tryAsStr();
-//                    ^^^^^^^^^^^^ Expected type to be "num?", but got "str?" instead
+//                 ^^^^^^^^^^^^ Expected type to be "num?", but got "str?" instead
 
 let tryStr: str? = j.tryAsBool();
-//                 ^^^^^^^^^^^^ Expected type to be "str?", but got "bool?" instead
+//                 ^^^^^^^^^^^^^ Expected type to be "str?", but got "bool?" instead
 
 let tryBool: bool? = j.tryAsNum();
-//                   ^^^^^^^^^^^^ Expected type to be "bool?", but got "num?" instead
\ No newline at end of file
+//                   ^^^^^^^^^^^^ Expected type to be "bool?", but got "num?" instead
+
+struct InnerStructyJson  {
+  good: bool;
+}
+
+struct StructyJson {
+  foo: str;
+  stuff: Array<num>;
+  maybe: InnerStructyJson;
+}
+
+struct StructyJsonMap {
+  stuff: Map<num>;
+}
+
+let notJsonMissingField: StructyJson = {
+  foo: "bar",
+  stuff: [],
+};
+//^ Missing required field "maybe" from "StructyJson"
+
+let notJsonBadNesting: StructyJson = {
+  foo: "bar",
+  stuff: [1, 2, 3],
+  maybe: {
+    good: 2,
+//        ^ Expected type to be "bool", but got "num" instead
+  }
+};
+
+let notJsonBadArray: StructyJson = {
+  foo: "bar",
+  stuff: [1, "hi", 3],
+//           ^^^^ Expected type to be "num", but got "str" instead
+  maybe: {
+    good: true,
+  }
+};
+
+
+let notJsonMap: StructyJsonMap = {
+  stuff: {
+    a: 1,
+    b: "",
+//     ^^ Expected type to be "num", but got "str" instead
+    c: 3,
+  },
+};
+
+let var mutableJson = {
+  foo: "bar",
+  stuff: [1, 2, 3],
+  maybe: {
+    good: true,
+  }
+};
+let mutableJsonStruct: StructyJson = mutableJson;
+//                                   ^^^^^^^^^^^ Expected type to be "StructyJson", but got "Json" instead
```

Comment:
How come this error disappears when the `var` keyword is removed?

I sort of get how this is coming about (if Json-typed variable is reassignable, then type inference becomes much more complicated, so the compiler stops trying to track the exact structure) -- but it feels weird because we're essentially saying that casting rules depend not only on types, but also whether we are casting from an lvalue or rvalue.

It's also weird because code like this seems like it should compile:
```js
let var mutableJson: StructyJson = {
  foo: "bar",
  stuff: [1, 2, 3],
  maybe: {
    good: true,
  }
};
```

If `mutableJson` is reassignable then `StructyJson` has to be on the right hand of the equal sign for the code to compile -- but if it's _not_ reassignable, then `StructyJson` can be on either side of the equal sign. This feels confusing.

I don't have a solution off the top of my head / not sure if this is a blocker, just sharing some raw thoughts here

@eladb @yoav-steinberg This seems like an interesting edge case

Reply:
> It's also weird because code like this seems like it should compile:

I agree, and it looks like this edge case is actually pretty easy to handle. Will add a fix.

As for the general case of `let mutableJsonStruct: StructyJson = mutableJson;`, it actually was pretty helpful when the displayed type included the known json info (the text that was removed in a recent thread). It told the user what our type system knew about that object, where just having `Json` told the user we had no info.

Reply:
An additional hint has been added to the error message to let the user know that they should use `.fromJson` because the data of the Json is not known.